src/persist-txn/src/txn_write.rs

```diff
@@ -0,0 +1,207 @@
+        commit_ts: u64,
+    ) -> Result<TxnApply, u64> {
+        // TODO(txn): Use ownership to disallow a double commit.
+        let mut txns_upper = txns::recent_upper(&mut handle.txns_write).await;
+
+        // Validate that the involved data shards are all registered. txns_upper
+        // only advances in the loop below, so we only have to check this once.
+        let () = handle.txns_cache.update_ge(txns_upper).await;
+        for (data_id, _) in self.writes.iter() {
+            let registered_before_commit_ts = handle
+                .txns_cache
+                .data_since(data_id)
+                .map_or(false, |x| x < commit_ts);
+            assert!(
+                registered_before_commit_ts,
+                "{} should be registered before commit at {}",
+                data_id, commit_ts
+            );
+        }
+
+        loop {
+            // txns_upper is the (inclusive) minimum timestamp at which we
+            // could possibly write. If our requested commit timestamp is before
+            // that, then it's no longer possible to write and the caller needs
+            // to decide what to do.
+            if commit_ts < txns_upper {
+                debug!("commit_at {} mismatch current={}", commit_ts, txns_upper);
+                return Err(txns_upper);
+            }
+            debug!(
+                "commit_at {}: [{}, {}) begin",
+                commit_ts,
+                txns_upper,
+                commit_ts + 1
+            );
+
+            let mut txn_batches = Vec::new();
+            let mut txns_updates = Vec::new();
+            for (data_id, updates) in self.writes.iter() {
+                let data_write = handle.datas.get_write(data_id).await;
+                // TODO(txn): Tighter lower bound?
+                let mut batch = data_write.builder(Antichain::from_elem(0));
+                for (k, d) in updates.iter() {
+                    batch.add(k, &(), &commit_ts, d).await.expect("valid usage");
+                }
+                let batch = batch
+                    .finish(Antichain::from_elem(commit_ts + 1))
+                    .await
+                    .expect("valid usage");
+                let batch = batch.into_writer_hollow_batch();
+                // TODO(txn): Proto not serde_json.
+                let batch_raw = serde_json::to_string(&batch).expect("valid json");
+                let batch = data_write.batch_from_hollow_batch(batch);
+                txn_batches.push(batch);
+                debug!(
+                    "wrote {:.9} batch {} len={}",
+                    data_id.to_string(),
+                    batch_raw.hashed(),
+                    updates.len()
+                );
+                txns_updates.push((*data_id, batch_raw));
+            }
+
+            let txns_updates = txns_updates
+                .iter()
+                .map(|(k, v)| ((k, v), commit_ts, 1))
+                .collect::<Vec<_>>();
+            let res = crate::small_caa(
```

Comment:
Is it possible that `txns_updates` may actually be quite large and we shouldn't be using `small_caa`?

Reply:
The same question applies to retracting data from the transaction shard in `apply_le`.

Reply:
i was thinking of "small" here being in some sense the control plane vs the data plan, so these two are exactly the sort of things i intended it for. plus, the "small"-ness is just some extra logging at debug-level, which is opt-in